ModelB_KD/KD/base_class.py

```diff
@@ -73,86 +73,96 @@ def __init__(
 
-    def train_teacher(
-        self,
-        epochs=20,
-        plot_losses=True,
-        save_model=True,
-        save_model_pth="./models/teacher.pt",
-    ):
-        """
-        Function that will be training the teacher
-
-        :param epochs (int): Number of epochs you want to train the teacher
-        :param plot_losses (bool): True if you want to plot the losses
-        :param save_model (bool): True if you want to save the teacher model
-        :param save_model_pth (str): Path where you want to store the teacher model
-        """
-        self.teacher_model.train()
-        loss_arr = []
-        length_of_dataset = len(self.train_loader.dataset)
-        best_acc = 0.0
-        self.best_teacher_model_weights = deepcopy(self.teacher_model.state_dict())
-
-        save_dir = os.path.dirname(save_model_pth)
-        if not os.path.exists(save_dir):
-            os.makedirs(save_dir)
-
-        print("Training Teacher... ")
-
-        for ep in range(epochs):
-            epoch_loss = 0.0
-            correct = 0
-            for (data, label) in self.train_loader:
-                data = data.to(self.device)
-                label = label.to(self.device)
-                out = self.teacher_model(data)
-
-                if isinstance(out, tuple):
-                    out = out[0]
-
-                pred = out.argmax(dim=1, keepdim=True)
-                correct += pred.eq(label.view_as(pred)).sum().item()
-
-                loss = self.ce_fn(out, label)
-
-                self.optimizer_teacher.zero_grad()
-                loss.backward()
-                self.optimizer_teacher.step()
-
-                epoch_loss += loss.item()
-
-            epoch_acc = correct / length_of_dataset
-
-            epoch_val_acc = self.evaluate(teacher=True)
-
-            if epoch_val_acc > best_acc:
-                best_acc = epoch_val_acc
-                self.best_teacher_model_weights = deepcopy(
-                    self.teacher_model.state_dict()
-                )
-
-            if self.log:
-                self.writer.add_scalar("Training loss/Teacher", epoch_loss, epochs)
-                self.writer.add_scalar("Training accuracy/Teacher", epoch_acc, epochs)
-                self.writer.add_scalar(
-                    "Validation accuracy/Teacher", epoch_val_acc, epochs
-                )
-
-            loss_arr.append(epoch_loss)
-            print(
-                "Epoch: {}, Loss: {}, Accuracy: {}".format(
-                    ep + 1, epoch_loss, epoch_acc
-                )
-            )
-
-            self.post_epoch_call(ep)
-
-        self.teacher_model.load_state_dict(self.best_teacher_model_weights)
-        if save_model:
-            torch.save(self.teacher_model.state_dict(), save_model_pth)
-        if plot_losses:
-            plt.plot(loss_arr)
+    def load_teacher(self, load_model_pth):
+        print("Loading Teacher")
+        self.teacher_model.load_state_dict(torch.load(load_model_pth))
```

Comment:
where do you get this from? have you verified it?

Reply:
have you verified it? If not, I believe it is wrong. Check my loading code for how to load a model

Reply:
I wrote this. 
https://pytorch.org/tutorials/beginner/saving_loading_models.html

It is verified. I have run it locally seeing a large change in accuracy in first couple data -> model is loaded -> verified.

Reply:
ok, so you are not using my model weights, right?

Reply:
If you are using your own model weights, then where is the weight for the large teacher model? Has it been trained?

If you are using my weights, this won't load.

Reply:
I am using the weights of the loaded teacher model. 

I am not sure why you say "it won't load" since you can see .weight parameter being initialized here as none zero values.  self.teacher_model is already initiated in the prepare function so why do you think it will not work?

<img width="1177" alt="Screenshot 2024-04-16 at 6 27 07 PM" src="https://github.com/QiyouChang/TrainTimeKD/assets/75822566/13745130-70c7-4bfd-86bf-16d553442396">

Reply:
This is how to load modelA weights. If you use your code, you won't be able to use the weights of modelA.https://github.com/QiyouChang/TrainTimeKD/blob/main/ModelA_Vit/vit_utils/model.py#L15

If you use weights trained by yourself, then we need to make sure that the pretrained model is large enough and good enough.